functional_tests/src/runtest/main.go

```diff
@@ -0,0 +1,191 @@
+	if *verbose {
+		info("Running command: %v", append([]string{name}, args...))
+	}
+	cmd := exec.Command(name, args...)
+	var stdoutBuffer bytes.Buffer
+	var stderrBuffer bytes.Buffer
+	cmd.Stdout = &stdoutBuffer
+	cmd.Stderr = &stderrBuffer
+	err = cmd.Run()
+	stdout = stdoutBuffer.String()
+	stderr = stderrBuffer.String()
+	if *verbose {
+		commandOutput("STDOUT", stdout)
+		commandOutput("STDERR", stderr)
+	}
+	return
+}
+
+func commandOutput(name string, content string) {
+	if len(content) <= 0 {
+		return
+	}
+
+	info("%s>>%s<<%s", name, content, name)
+}
+
+func doSetup(suite Suite) {
+	info(">>> Setting up...")
+	for _, setup := range suite.Setup {
+		err, _, _ := runCommand(setup.Command[:1][0], setup.Command[1:]...)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+}
+
+func doTeardown(suite Suite) {
+	info(">>> Tearing down...")
+	for _, teardown := range suite.Teardown {
+		err, _, _ := runCommand(teardown.Command[:1][0], teardown.Command[1:]...)
+		if err != nil {
+			info(" > Warning: Teardown command failed: %s", err)
+		}
+	}
+}
+
+func doRunTests(suite Suite) func() {
+	info(">>> Testing...")
+	results := make(map[int]string)
+	passing := make(map[int]bool)
+	for index, test := range suite.Tests {
+		doOneTest(index, test, suite, results, passing)
+	}
+
+	report := func() {
+		info(">>> Summary:")
+		for index := range suite.Tests {
+			info(" > %s", results[index])
+		}
+
+		for index := range suite.Tests {
+			if !passing[index] {
+				log.Fatalf("Some test has failed.")
+			}
+		}
+	}
+	return report
+}
+
+func doOneTest(index int, test Test, suite Suite, results map[int]string, passing map[int]bool) {
+	var name string
+	var msg string
+	var result string
+
+	if len(test.Name) > 0 {
+		name = fmt.Sprintf("test-%d (%s)", index, test.Name)
+	} else {
+		name = fmt.Sprintf("test-%d", index)
+	}
+	info(" > %s", name)
+
+	recordResult := func(b *string) {
+		if *b == PASSED {
+			results[index] = fmt.Sprintf("%s: %s", name, PASSED)
+			passing[index] = true
+		} else {
+			results[index] = fmt.Sprintf("%s: %s", name, FAILED)
+			passing[index] = false
+		}
+		info(" %s", *b)
+	}
+	defer recordResult(&result)
+
+	args := append([]string{"exec", suite.Target}, test.Command...)
+	err, stdout, stderr := runCommand("docker", args...)
+
+	if err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				result = fmt.Sprintf("%s: Exit status %d", FAILED, status.ExitStatus())
+			}
+		} else {
+			result = fmt.Sprintf("%s: Encountered error: %v", FAILED, err)
+		}
+		return
+	}
+
+	msg = doStringAssert(stdout, test.Expect.Stdout)
+	if len(msg) > 0 {
+		result = fmt.Sprintf("%s: stdout assertion failure\n%s", FAILED, msg)
+		return
+	}
+	msg = doStringAssert(stderr, test.Expect.Stderr)
+	if len(msg) > 0 {
+		result = fmt.Sprintf("%s: stderr assertion failure\n%s", FAILED, msg)
+		return
+	}
+
+	result = PASSED
+}
+
+func doStringAssert(value string, rule StringAssert) string {
```

Comment:
I think might be a little more clear if you use regexes here. Then you could get rid of the different fields in the yaml spec (`exactly` vs. `equals`) and just have the user specify the regex to match. WDYT?